### PR TITLE
Fix OpenAPI generation on Windows

### LINKED ISSUE #16759: Openapi code generation failed on zeebe-client-java
**Describe the bug**

 Failed to execute goal org.openapitools:openapi-generator-maven-plugin:7.3.0:generate (rest) on project zeebe-client-java: Code generation failed.

**To Reproduce**

Run `mvn clean package -DskipTests`  in this project.

**Expected behavior**

Run mvn cmd sucessful

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
[INFO] --- openapi-generator:7.3.0:generate (rest) @ zeebe-client-java ---
[ERROR] Error resolving #/components/responses/TopologyResponse
java.net.URISyntaxException: Illegal character in opaque part at index 2: D:\github\skayliu\zeebe/zeebe/gateway-protocol/src/main/proto/rest-api.yaml
    at java.net.URI$Parser.fail (URI.java:2995)
    at java.net.URI$Parser.checkChars (URI.java:3166)
    at java.net.URI$Parser.parse (URI.java:3202)
    at java.net.URI.<init> (URI.java:645)
    at io.swagger.v3.parser.reference.Visitor.readURI (Visitor.java:79)
    at io.swagger.v3.parser.reference.ReferenceVisitor.parse (ReferenceVisitor.java:305)
    at io.swagger.v3.parser.reference.ReferenceVisitor.toReference (ReferenceVisitor.java:74)
    at io.swagger.v3.parser.reference.ReferenceVisitor.resolveRef (ReferenceVisitor.java:205)
    at io.swagger.v3.parser.reference.ReferenceVisitor.visitResponse (ReferenceVisitor.java:152)
    at io.swagger.v3.parser.reference.OpenAPI31Traverser.traverseResponse (OpenAPI31Traverser.java:289)
    at io.swagger.v3.parser.reference.OpenAPI31Traverser.traverseMap (OpenAPI31Traverser.java:933)
    at io.swagger.v3.parser.reference.OpenAPI31Traverser.traverseResponses (OpenAPI31Traverser.java:270)
    at io.swagger.v3.parser.reference.OpenAPI31Traverser.traverseOperation (OpenAPI31Traverser.java:237)
    at io.swagger.v3.parser.reference.OpenAPI31Traverser.traversePathItem (OpenAPI31Traverser.java:394)
    at io.swagger.v3.parser.reference.OpenAPI31Traverser.traverseMap (OpenAPI31Traverser.java:933)
    at io.swagger.v3.parser.reference.OpenAPI31Traverser.traversePaths (OpenAPI31Traverser.java:197)
    at io.swagger.v3.parser.reference.OpenAPI31Traverser.traverseOpenApi (OpenAPI31Traverser.java:124)
    at io.swagger.v3.parser.reference.OpenAPI31Traverser.traverse (OpenAPI31Traverser.java:65)
    at io.swagger.v3.parser.reference.OpenAPIDereferencer31.dereference (OpenAPIDereferencer31.java:74)
    at io.swagger.v3.parser.OpenAPIV3Parser.resolve (OpenAPIV3Parser.java:225)
    at io.swagger.v3.parser.OpenAPIV3Parser.readContents (OpenAPIV3Parser.java:181)
    at io.swagger.v3.parser.OpenAPIV3Parser.readLocation (OpenAPIV3Parser.java:97)
    at io.swagger.parser.OpenAPIParser.readLocation (OpenAPIParser.java:16)
    at org.openapitools.codegen.config.CodegenConfigurator.toContext (CodegenConfigurator.java:686)
    at org.openapitools.codegen.config.CodegenConfigurator.toClientOptInput (CodegenConfigurator.java:744)
    at org.openapitools.codegen.plugin.CodeGenMojo.execute (CodeGenMojo.java:924)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:283)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:226)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:407)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:348)
[ERROR]
org.openapitools.codegen.SpecValidationException: There were issues with the specification. The option can be disabled via validateSpec (Maven/Gradle) or --skip-validate-spec (CLI).
 | Error count: 1, Warning count: 2
Errors:
        -Illegal character in opaque part at index 2: D:\github\skayliu\zeebe/zeebe/gateway-protocol/src/main/proto/rest-api.yaml
Warnings:
        -Illegal character in opaque part at index 2: D:\github\skayliu\zeebe/zeebe/gateway-protocol/src/main/proto/rest-api.yaml

    at org.openapitools.codegen.config.CodegenConfigurator.toContext (CodegenConfigurator.java:717)
    at org.openapitools.codegen.config.CodegenConfigurator.toClientOptInput (CodegenConfigurator.java:744)
    at org.openapitools.codegen.plugin.CodeGenMojo.execute (CodeGenMojo.java:924)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:283)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:226)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:407)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:348)
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Zeebe Root 8.5.0-SNAPSHOT:
[INFO]
[INFO] Zeebe BOM .......................................... SUCCESS [  1.340 s]
[INFO] Zeebe Build Tools .................................. SUCCESS [  1.497 s]
[INFO] Zeebe Parent ....................................... SUCCESS [  2.413 s]
[INFO] Zeebe Util ......................................... SUCCESS [  4.952 s]
[INFO] Zeebe Atomix Parent Pom ............................ SUCCESS [  1.273 s]
[INFO] Zeebe Atomix Utilities ............................. SUCCESS [  2.588 s]
[INFO] Zeebe Scheduler .................................... SUCCESS [  4.105 s]
[INFO] Zeebe Protocol ..................................... SUCCESS [  6.672 s]
[INFO] Zeebe Exporter API ................................. SUCCESS [  1.487 s]
[INFO] Zeebe BPMN model API ............................... SUCCESS [ 10.396 s]
[INFO] Zeebe Gateway Protocol ............................. SUCCESS [  1.692 s]
[INFO] Zeebe Gateway Protocol Implementation .............. SUCCESS [  5.027 s]
[INFO] Zeebe Client Java .................................. FAILURE [  9.316 s]
[INFO] Zeebe Test Util .................................... SKIPPED
[INFO] Zeebe Snapshots .................................... SKIPPED
[INFO] Zeebe Journal ...................................... SKIPPED
[INFO] Zeebe Atomix Cluster ............................... SKIPPED
[INFO] Zeebe DB ........................................... SKIPPED
[INFO] Zeebe Msgpack Core ................................. SKIPPED
[INFO] Zeebe Msgpack Value ................................ SKIPPED
[INFO] Zeebe Auth ......................................... SKIPPED
[INFO] Zeebe Protocol Implementation ...................... SKIPPED
[INFO] Zeebe Logstreams ................................... SKIPPED
[INFO] Zeebe Stream Platform .............................. SKIPPED
[INFO] Zeebe Transport .................................... SKIPPED
[INFO] Zeebe Cluster Topology ............................. SKIPPED
[INFO] Zeebe Protocol AssertJ Assertions .................. SKIPPED
[INFO] Zeebe Protocol Test Util ........................... SKIPPED
[INFO] Zeebe Broker Client ................................ SKIPPED
[INFO] Zeebe Gateway ...................................... SKIPPED
[INFO] Zeebe FEEL Integration ............................. SKIPPED
[INFO] Zeebe Expression Language .......................... SKIPPED
[INFO] Zeebe DMN Engine ................................... SKIPPED
[INFO] Zeebe Workflow Engine .............................. SKIPPED
[INFO] Zeebe Backup ....................................... SKIPPED
[INFO] Zeebe Backup Store Testkit ......................... SKIPPED
[INFO] Zeebe Backup Store for S3 .......................... SKIPPED
[INFO] Zeebe Backup Store Common .......................... SKIPPED
[INFO] Zeebe Backup Store for GCS ......................... SKIPPED
[INFO] Zeebe Backup Store for Azure ....................... SKIPPED
[INFO] Zeebe Exporter Test Harness ........................ SKIPPED
[INFO] Zeebe Broker ....................................... SKIPPED
[INFO] Zeebe Backup Restore ............................... SKIPPED
[INFO] Zeebe Gateway REST API server ...................... SKIPPED
[INFO] Zeebe Protocol Jackson ............................. SKIPPED
[INFO] Zeebe Elasticsearch Exporter ....................... SKIPPED
[INFO] Zeebe Opensearch Exporter .......................... SKIPPED
[INFO] Zeebe Distribution ................................. SKIPPED
[INFO] Zeebe QA ........................................... SKIPPED
[INFO] Zeebe QA Utilities ................................. SKIPPED
[INFO] Zeebe QA Integration Tests ......................... SKIPPED
[INFO] Zeebe Benchmark .................................... SKIPPED
[INFO] Zeebe Root ......................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  53.875 s
[INFO] Finished at: 2024-03-06T18:36:48+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.openapitools:openapi-generator-maven-plugin:7.3.0:generate (rest) on project zeebe-client-java: Code generation failed. See above for the full exception. -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :zeebe-client-java

```

</p>
</details>

**Environment:**
- OS: Windows 10 x64 
- Zeebe Version: Master branch
- Configuration: None

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR fixes the OpenAPI generator issues on Windows by downgrading the spec from 3.1.0 to 3.0.3. This also aligns it with the rest of C8, as everyone is on 3.0.x.

Doing so re-introduced the `JsonNullable` dependency, which then needs to be added to the build, and its serialization module also added to any `ObjectMapper` that will be used for serialization (both client and server).

Since OpenAPI < 3.1.0 does not allow adding _any_ properties next to a `$ref`, in order to keep the description I had to be a bit creative.

The `UserTaskUpdateRequest` changeset property now "inherits" from the changeset property. This is not true inheritance for the generated model, but more composition. So it has the same properties, its own description, etc., but does not directly inherit from the `Changeset` class. This means it loses the inheritance to `HashMap`. Unfortunately, adding `additionalProperty` to it failed - while it would now inherit from `HashMap`, it failed to import the class in the generated code. Since additional properties are effectively ignored for this, however, I figured it was fine as is.

I also added the Java 8 modules for Jackson on both the server and client already. I would recommend making use of them for consistent parsing between both.

## Related issues

closes #16759 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: zeebe/clients/java/pom.xml
Comment (npepinpe) [inline zeebe/clients/java/pom.xml:296]: :shrug: Not sure why this was there

File: zeebe/clients/java/pom.xml
Comment (npepinpe) [inline zeebe/clients/java/pom.xml:291]: :thought_balloon: I would strongly recommend using an appropriate type here and not just strings to avoid mismatches in general :shrug:

**Review by npepinpe — COMMENTED** [review]
There's a big downside of not having a single module for the code generation, and that's there's a lot of shared options you need to update on both sides each time. It might be good to have that somewhere, but I didn't give it so much thought. Maybe it's something you only need to set up once, I guess...

**Comment by npepinpe** [comment]
@tmetzke - sorry this kind of became bigger than expected. I highlighted a few reasons (serialization can be brittle between client/server if we're not setting up Jackson the same way, the generator kind of sucks in general...), hope that's alright =/

File: zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/command/UpdateUserTaskCommandImpl.java
Comment (tmetzke) [inline zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/command/UpdateUserTaskCommandImpl.java:77]: This won't work, unfortunately. This is also why the QA tests are failing. The changeset is handled as a Java Map when serialized. Thus, only the entries in the map itself are serialized. Additional attributes like the `dueDate` are not. See https://github.com/camunda/zeebe/pull/16676#issuecomment-1976393413.

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml
Comment (tmetzke) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:69]: 👍 That's a neat trick

File: zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/command/UpdateUserTaskCommandImpl.java
Comment (npepinpe) [reply zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/command/UpdateUserTaskCommandImpl.java:77]: Tests are passing now though (locally anyway)

Comment (npepinpe) [reply zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/command/UpdateUserTaskCommandImpl.java:77]: AFAIK the property is empty, but it's non null, so we still serialize something in the end.

**Comment by npepinpe** [comment]
This was more work than I expected, and it's making me dislike the generator more and more :smile:

File: zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/command/UpdateUserTaskCommandImpl.java
Comment (tmetzke) [reply zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/command/UpdateUserTaskCommandImpl.java:77]: This a POJO class in the client now due to the changes and not a subtype of `HashMap` anymore, right? Than those changes should be fine, yeah, you're right 👍

**Comment by npepinpe** [comment]
It also makes me appreciate how code we had it with gRPC in terms of generated code.

File: zeebe/clients/java/pom.xml
Comment (tmetzke) [reply zeebe/clients/java/pom.xml:292]: 🤷

File: zeebe/clients/java/pom.xml
Comment (tmetzke) [reply zeebe/clients/java/pom.xml:287]: Can't test the impact locally right now, unfortunately. If you think we should remove this from the client and it still works, I'm fine with that.

**Comment by npepinpe** [comment]
Last commits fix some Go test issues because the serialized duration format changed, and Go wasn't expecting it :D

**Comment by npepinpe** [comment]
Since changes to the main distro may break other stuff, I'm gonna run a benchmark pre-emptively.

**Comment by npepinpe** [comment]
@tmetzke - please take over tomorrow morning to merge this. I've already notified our release manager. If you see it keeps getting out of control, then we can postpone this, but it would be a shame as it may end up breaking other things (specifically the serialization) later.

**Comment by tmetzke** [comment]
> ### Setup
> Deployed to **measurement-8525971808**
> 
> ```yaml
> camunda-platform:
>   zeebe:
>     image:
>       repository: gcr.io/zeebe-io/zeebe
>       tag: np-16759-fix-benchmark-42718db
>   zeebe-gateway:
>     image:
>       repository: gcr.io/zeebe-io/zeebe
>       tag: np-16759-fix-benchmark-42718db
> global:
>   image:
>     tag: np-16759-fix-benchmark-42718db
> ```
> 
> ### Measurement before
> **Process Instance Execution Time**: p99=0.981 p90=0.805 p50=0.442 **Throughput**: 51.005 PI/s [Grafana](https://grafana.dev.zeebe.io/d/zeebe-dashboard/zeebe?orgId=1&var-namespace=measurement-8525971808&from=1712074715219&to=1712074817101)
> 
> ### Chaos injection
> Deployed chaos network-latency-5
> 
> ### Measurement after
> **Process Instance Execution Time**: p99=0.590 p90=0.207 p50=0.068 **Throughput**: 150.260 PI/s [Grafana](https://grafana.dev.zeebe.io/d/zeebe-dashboard/zeebe?orgId=1&var-namespace=measurement-8525971808&from=1712074947030&to=1712075822539)
> 
> ### Details
> See https://github.com/camunda/zeebe/actions/runs/8525971808

Wait, did the throughput just triple? 😁

File: zeebe/clients/java/pom.xml
Comment (npepinpe) [reply zeebe/clients/java/pom.xml:287]: I just added it for consistency, as otherwise the client may not send it as a string. I think it's OK, it's more for the Java side - we're enforcing RFC3339, but not using the type system to do that :shrug:

Comment (tmetzke) [reply zeebe/clients/java/pom.xml:287]: OK, I see. I'll give this another look locally now to double-check. I get your point 👍

Comment (tmetzke) [reply zeebe/clients/java/pom.xml:287]: OK, so, this is actually necessary now, thanks for adding it, @npepinpe. Otherwise, we cannot reset the date values. We need to send empty strings to the server to reset the values. If the client uses proper `OffsetDateTime`s, you cannot send empty strings. This wasn't necessary before because we used the free form attributes instead of the typed attributes of the request object.

**Review by tmetzke — APPROVED** [review]
👍 Thanks again, @npepinpe, and sorry that this became way more than anticipated.
❓ If you still read this before we merge, I was wondering if we actually need the time modules still although we're using plain strings on both ends for the dates (client and server).

**Comment by npepinpe** [comment]
We don't need them on the client, but TIL that by default Spring was loading them, and if we now provide our own Jackson customizer which calls `modules` (which, as per their doc, is _not additive, but overwrites previous values_), then we were losing them, resulting in other things failing. Thanks, Spring magic...

**Comment by npepinpe** [comment]
By the way, there's a bunch of Operate tests failing, but beats me if it's relevant or not...

**Pull request merged**
